numba/typing/arraydecl.py

```diff
@@ -292,9 +292,39 @@ def _resolve_real_imag(self, ary, attr):
 
     @bound_function("array.transpose")
     def resolve_transpose(self, ary, args, kws):
-        assert not args
+        def sentry_shape_scalar(ty):
+            if ty in types.number_domain:
+                # Guard against non integer type
+                if not isinstance(ty, types.Integer):
+                    raise TypeError("transpose() arg cannot be {0}".format(ty))
```

Comment:
A test should be written to make sure these guards (and [here](https://github.com/numba/numba/pull/2762/files#diff-7efa237a0406240ad8523ad2443fa960R324)) are triggered.

Reply:
Added a test for this case.  I could not find a test that tests the vararg case in the existing code.  For example, although array.reshape has vararg form, it is not tested.  I think this is because typeof() does not support vararg, therefore there are no such tests.  Please correct me if I'm wrong.

Reply:
That does sound feasible and probably needs fixing. Thanks for adding the other tests.